libs/runtime/ebpf_object.c

```diff
@@ -154,26 +157,48 @@ ebpf_object_tracking_initiate()
     memset(_ebpf_object_reference_history, 0, sizeof(_ebpf_object_reference_history));
     _ebpf_object_reference_history_index = 0;
 
+    cxplat_initialize_rundown_protection(&_ebpf_object_rundown_ref);
+
     return ebpf_hash_table_create(&_ebpf_id_table, &options);
 }
 
 void
 ebpf_object_tracking_terminate()
 {
-    ebpf_assert(!_ebpf_id_table || ebpf_hash_table_key_count(_ebpf_id_table) == 0 || ebpf_fault_injection_is_enabled());
+    if (!_ebpf_id_table) {
+        return;
+    }
+
+    cxplat_wait_for_rundown_protection_release(&_ebpf_object_rundown_ref);
+
     ebpf_hash_table_destroy(_ebpf_id_table);
     _ebpf_id_table = NULL;
 }
 
+static void
+_ebpf_object_epoch_free(_Inout_ void* context)
+{
+    ebpf_core_object_t* object = (ebpf_core_object_t*)context;
+    EBPF_LOG_MESSAGE_POINTER_ENUM(
+        EBPF_TRACELOG_LEVEL_VERBOSE, EBPF_TRACELOG_KEYWORD_BASE, "eBPF object terminated", object, object->type);
+
+    object->base.marker = ~object->base.marker;
+    object->free_function(object);
+    cxplat_release_rundown_protection(&_ebpf_object_rundown_ref);
```

Comment:
In general, at what point do we need to acquire and release the rundown protection? Is that done right after acquiring a ref to the object, and then right after releasing it?

Reply:
Implied assumption in the code is that when ebpf_object_tracking_terminate returns then all objects have been freed. Given that they freed on epoch end, there is a need to separately track when all the objects have been freed.